content/blog/2023-12-11-polyhedral/index.md

```diff
@@ -0,0 +1,134 @@
+
+To obtain MLIR from benchmarks written in C, I used the existing [Polygeist](https://polygeist.llvm.org/) frontend configured to emit the `scf` dialect.
+
+## Implementation
+
+I was less than successful (to put it charitably) in realizing the goal of an end-to-end optimization pass. The current implementation performs, I believe, the bulk of the translation into the polyhedral model, but notably lacks any facilities for obtaining transformed code out of the model, and the computation of the legal transformation space is incomplete.
+
+The implementation currently performs the following analyses, each of which is described in detail below.
+
+- Detect static control parts.
+- Construct the polyhedral representation of each statement's iteration domain.
+- Construct the dependence polyhedra for a given pair of memory references.
+
+### Building Static Control Parts
+
+Control flow in the polyhedral model comprises a restricted class of loop nests whose bounds can be specified as a set of affine inequalities, such that the iteration domain forms a convex polyhedron. The class of program fragments amenable to polyhedral representation (that is, with convex polyhedral iteration domain) is said to have *static control*. A maximal region with static control is referred to as a *static control part*.
+
+The first step in our analysis is thus the identification of these program regions. We follow the basic algorithm given by Bastoul:
+
+1. Walk the IR, identifying operations with static control: loops with affine bounds and unit step, conditionals whose conditions are affine inequalities, and memory references with affine subscripts. In all of the these cases, affine functions may depend only on surrounding induction variables, constants, and parameters not mutated within the SCoP.
+2. Merge consecutive operations with static control into static control parts.
+
+The result is a list of the SCoPs within a given function. Implementing this analysis in MLIR is relatively straightforward due to the IR's explicitly nesting regions.
+
+### Building Iteration Domains
+
+Having identified regions with polyhedral domain, the next step is to construct those domains explicitly in a format that can be manipulated by integer programming tools. We use [isl](https://libisl.sourceforge.io/) to build sets from affine constraints and discharge queries such as emptiness checking.
+
+For a given operation, we begin by determining the space in which its iteration domain should be constructed. This determination depends on the SCoP in which the operation resides: loop induction variables within the SCoP correspond to dimensions of the space, while other variables correspond to parameters.
```

Comment:
I think this is just my ignorance about the polyhedral model showing, but I don't quite grok what "space" and "dimensions" mean concretely here. Maybe it would be helpful to add a little context-setting above about what isl defines a "domain" to be? What are all the ingredients required to create an iteration domain?